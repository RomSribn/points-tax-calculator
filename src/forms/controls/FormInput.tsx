import React from 'react';
import classNames from 'classnames';
import get from 'lodash.get';
import { DeepMap, FieldError, UseFormRegister, Path, FieldValues } from 'react-hook-form';
import { ErrorMessage } from '@hookform/error-message';

import { Input, InputProps, ErrorMessage as FormErrorMessage } from '../_elements';

export type FormInputProps<TFormValues extends FieldValues> = {
  name: Path<TFormValues>;
  register?: UseFormRegister<TFormValues>;
  errors?: Partial<DeepMap<TFormValues, FieldError>>;
} & Omit<InputProps, 'name'>;

const FormInput = <TFormValues extends Record<string, unknown>>({
  name,
  register,
  errors,
  className,
  ...props
}: FormInputProps<TFormValues>): JSX.Element => {
  // If the name is in a FieldArray, it will be 'fields.index.fieldName' and errors[name] won't return anything, so we are using lodash get
  const errorMessages = get(errors, name);
  const hasError = !!(errors && errorMessages);

  return (
    <div className={classNames(className)} aria-live="polite" data-testid="form-input">
      <Input
        name={name}
        aria-invalid={hasError}
        className={classNames({
          'transition-colors focus:outline-none focus:ring-2 focus:ring-opacity-50 border-red-600 hover:border-red-600 focus:border-red-600 focus:ring-red-600':
            hasError
        })}
        {...props}
        {...(register && register(name))}
      />
      <ErrorMessage
        errors={errors}
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        name={name as any}
        render={({ message }) => <FormErrorMessage className="mt-1 text-danger">{message}</FormErrorMessage>}
      />
    </div>
  );
};

export { FormInput };
