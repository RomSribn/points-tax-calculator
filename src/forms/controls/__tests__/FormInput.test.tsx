import React from 'react';
import { render, screen } from '@testing-library/react';

import { FormInput, FormInputProps } from '../FormInput';

describe('FormInput component', () => {
  test('renders input and error message elements', () => {
    const name = 'input-name';
    const inputId = 'input-id';
    const inputLabel = 'Input Label';
    const customClass = 'custom-class';
    const placeHolder = 'Enter a value';
    const register = jest.fn();
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    const errors: FormInputProps<TFormValues>['errors'] = {
      [name]: {
        message: 'Input is required'
      }
    };

    render(
      <FormInput
        name={name}
        register={register}
        errors={errors}
        id={inputId}
        label={inputLabel}
        className={customClass}
        placeholder={placeHolder}
      />
    );

    const formInputContainer = screen.getByTestId('form-input');
    const inputChildElement = screen.getByLabelText(inputLabel) as HTMLInputElement;
    const errorMessageChildElement = screen.getByText(errors[name].message);

    expect(inputChildElement).toBeInTheDocument();
    expect(inputChildElement.id).toBe(inputId);
    expect(inputChildElement.name).toBe(name);
    expect(inputChildElement.placeholder).toBe(placeHolder);
    expect(formInputContainer.className).toContain(customClass);

    expect(errorMessageChildElement).toBeInTheDocument();
    expect(formInputContainer).toMatchSnapshot();
  });
});
