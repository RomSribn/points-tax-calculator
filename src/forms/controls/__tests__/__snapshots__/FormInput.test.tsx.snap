// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormInput component renders input and error message elements 1`] = `
<div
  aria-live="polite"
  class="custom-class"
  data-testid="form-input"
>
  <div
    class="form-group"
  >
    <input
      aria-invalid="true"
      aria-label="Input Label"
      class="form-control transition-colors focus:outline-none focus:ring-2 focus:ring-opacity-50 border-red-600 hover:border-red-600 focus:border-red-600 focus:ring-red-600"
      id="input-id"
      name="input-name"
      placeholder="Enter a value"
      type="text"
    />
  </div>
  <p
    class="font-serif text-sm text-left block text-red-600 mt-1 text-danger"
    data-testid="error-message"
  >
    Input is required
  </p>
</div>
`;
